docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md

```diff
@@ -0,0 +1,213 @@
+    - the public `ssh` key of the user is set under `spec.sshPublicKey`. The key that should be used needs to be configured beforehand by the user
+    - The targeted shoot is set under `spec.shootRef`
+2. GAPI Admission Control for the `Bastion` resource in the garden cluster
+    - Mutating Webhook
+        - on creation, sets `metadata.annotations["operations.gardener.cloud/created-by"]` according to the user that created the resource
+        - when `operations.gardener.cloud/operation: keepalive` is set, calculates new `status.expirationTimestamp`
+3. `gardenlet`
+    - Watches `Bastion` resource for own seed under api group `operations.gardener.cloud` in the garden cluster
+    - Creates `Bastion` custom resource under api group `extensions.gardener.cloud/v1alpha1` in the seed cluster
+4. `GCM`:
+    - During reconcile of the `Bastion` resource:
+      - according to `spec.shootRef`, sets the `status.seedName`
+      - according to `spec.shootRef`, sets the `status.providerType`
+5. Gardener extension provider <infra> / Bastion Controller on Seed:
+    - With own `Bastion` Custom Resource Definition in the seed under the api group `extensions.gardener.cloud/v1alpha1`
+    - Watches `Bastion` custom resources that are created by the `gardenlet` in the seed
+    - Controller reads `cloudprovider` credentials from seed-shoot namespace
+    - Deploy infrastructure resources
+        - Bastion VM. User data similar to https://github.com/gardener/gardenctl/blob/1e3e5fa1d5603e2161f45046ba7c6b5b4107369e/pkg/cmd/ssh.go#L160-L171. Writes `spec.sshPublicKey` into `authorized_keys` file.
+        - attaches public IP, creates security group, firewall rules, etc.
+    - Updates status of `Bastion` resource:
+        - With bastion IP under `status.ingress.ip` or hostname under `status.ingress.hostname`
+        - Updates the `status.lastOperation` with the status of the last reconcile operation
+6. `gardenlet`
+    - Syncs back the `status.ingress` and `status.conditions` of the `Bastion` resource in the seed to the garden cluster in case it changed
+7. `gardenctl`
+    - initiates `ssh` session once `status.operation['BastionReady']` is true of the `Bastion` resource in the garden cluster
+        - locates private `ssh` key matching `spec["sshPublicKey"]` which was configured beforehand by the user
+        - reads bastion IP (`status.ingress.ip`) or hostname (`status.ingress.hostname`)
+        - reads the private key from the `ssh` key pair for the shoot node
+        - opens `ssh` connection to the bastion and from there to the respective shoot node
+    - runs heartbeat in parallel as long as the `ssh` session is open by annotating the `Bastion` resource with `operations.gardener.cloud/operation: keepalive`
```

Comment:
When I SSH'ed into the bastion VM, how do I get to a worker node from here? Do I need to download the SSH key of the worker nodes myself and copy it to the VM, or will something take care automatically and provide the key?

Reply:
see my comment here https://github.com/gardener/gardener/pull/3802#discussion_r614336968

there should be no need to copy it to the vm I presume, see `proxycommand` https://github.com/gardener/gardenctl/blob/master/pkg/cmd/ssh_alicloud.go#L113 as rough example

Reply:
Hm, I didn't get it, can you explain again?

Reply:
does the example here help https://github.com/gardener/gardener/pull/3802#discussion_r614623942 ?

Reply:
Yep, I answered in https://github.com/gardener/gardener/pull/3802#discussion_r614676673, thank you, let's continue there (resolving this comment).